File: packages/mesh-runtime/test/common/renderer.mjs
// The reference renderer (D10), in Node: the Rust tests' renderer
// (`crates/mesh-runtime/tests/common/renderer.rs`), with the same printed
// form, so both produce the same committed files. It draws what it's
// given and computes nothing: prop values print as the JSON they are,
// and text runs as their strings.
import { byCodePoint, canonical } from "./json.mjs";

/** The tree as indented, HTML-like text; keys aren't printed. */
export function print(tree) {
  const out = [];
  node(tree.root, 0, out);
  return out.join("");
}

function node(n, depth, out) {
  const indent = "  ".repeat(depth);
  let tag = `${indent}<${n.component}`;
  for (const name of Object.keys(n.props).sort(byCodePoint)) {
    tag += ` ${name}=${canonical(n.props[name])}`;
  }
  for (const event of Object.keys(n.events).sort(byCodePoint)) {
    tag += ` on.${event}`;
  }
  if (n.children.length === 0) {
    out.push(`${tag} />\n`);
    return;
  }
  out.push(`${tag}>\n`);
  for (const child of n.children) {
    if (child.type === "node") {
      node(child, depth + 1, out);
    } else {
      out.push(`${indent}  ${JSON.stringify(child.text)}\n`);
    }
  }
  out.push(`${indent}</${n.component}>\n`);
}

/** Every node and text run, by key: where it is, and its props or text. */
function parts(n, at, out) {
  out.set(n.key, { kind: "node", at, props: n.props });
  n.children.forEach((child, index) => {
    if (child.type === "node") {
      parts(child, `${at}/${child.component}[${index}]`, out);
    } else {
      out.set(child.key, { kind: "text", at: `${at}/text[${index}]`, text: child.text });
    }
  });
}

/**
 * Every prop and text run whose value changed from `old` to `next`,
 * matched by key, one per line. Throws if the trees' keys differ: two
 * renders of one program always have the same keys.
 */
export function compare(old, next) {
  const before = new Map();
  const after = new Map();
  parts(old.root, old.root.component, before);
  parts(next.root, next.root.component, after);
  const keys = (map) => [...map.keys()].sort(byCodePoint);
  if (canonical(keys(before)) !== canonical(keys(after))) {
    throw new Error("two renders of one program have the same keys");
  }
  const changes = [];
  for (const key of keys(after)) {
    const a = before.get(key);
    const b = after.get(key);
    if (a.kind !== b.kind) throw new Error("a key changed kind");
    if (b.kind === "node") {
      const names = [...new Set([...Object.keys(a.props), ...Object.keys(b.props)])].sort(byCodePoint);
      for (const name of names) {
        const show = (props) => (name in props ? canonical(props[name]) : "(absent)");
        if (show(a.props) !== show(b.props)) {
          changes.push([b.at, `${b.at} ${name}: ${show(a.props)} -> ${show(b.props)}`]);
        }
      }
    } else if (a.text !== b.text) {
      changes.push([b.at, `${b.at}: ${JSON.stringify(a.text)} -> ${JSON.stringify(b.text)}`]);
    }
  }
  changes.sort((x, y) => byCodePoint(x[0], y[0]) || byCodePoint(x[1], y[1]));
  return changes.map(([, line]) => `${line}\n`).join("");
}

File: crates/mesh-runtime/tests/common/renderer.rs
//! The reference renderer (D10), test-only: prints a render tree as
//! HTML-like text, so an expected rendering can be a committed file, and
//! compares two trees key by key, as a renderer reconciling in place
//! would (D6).
//!
//! It draws what it's given and computes nothing: prop values are printed
//! as the JSON they are, and text runs as their strings.

use mesh_runtime::{Node, Tree, TreeChild};
use std::collections::BTreeMap;
use std::fmt::Write;

/// The tree as indented, HTML-like text: one node per line, props in
/// code-point order as `name=<json>`, events as `on.<event>`, text runs
/// quoted. Keys aren't printed: they're opaque, and change with the
/// program.
pub fn print(tree: &Tree) -> String {
    let mut out = String::new();
    node(&tree.root, 0, &mut out);
    out
}

fn node(node: &Node, depth: usize, out: &mut String) {
    let indent = "  ".repeat(depth);
    let mut tag = format!("{indent}<{}", node.component);
    for (name, value) in &node.props {
        write!(tag, " {name}={value}").unwrap();
    }
    for event in node.events.keys() {
        write!(tag, " on.{event}").unwrap();
    }
    if node.children.is_empty() {
        out.push_str(&tag);
        out.push_str(" />\n");
        return;
    }
    out.push_str(&tag);
    out.push_str(">\n");
    for child in &node.children {
        match child {
            TreeChild::Node(child) => self::node(child, depth + 1, out),
            TreeChild::Text { text, .. } => {
                writeln!(out, "{indent}  {}", serde_json::to_string(text).unwrap()).unwrap();
            }
        }
    }
    writeln!(out, "{indent}</{}>", node.component).unwrap();
}

/// What a part of the tree is, by key: where it is (for people), and its
/// props or text (for comparing).
enum Part {
    Node {
        at: String,
        props: BTreeMap<String, serde_json::Value>,
    },
    Text {
        at: String,
        text: String,
    },
}

fn parts(node: &Node, at: String, out: &mut BTreeMap<String, Part>) {
    out.insert(
        node.key.clone(),
        Part::Node {
            at: at.clone(),
            props: node.props.clone(),
        },
    );
    for (index, child) in node.children.iter().enumerate() {
        match child {
            TreeChild::Node(child) => {
                parts(child, format!("{at}/{}[{index}]", child.component), out)
            }
            TreeChild::Text { key, text } => {
                out.insert(
                    key.clone(),
                    Part::Text {
                        at: format!("{at}/text[{index}]"),
                        text: text.clone(),
                    },
                );
            }
        }
    }
}

/// Every prop and text run whose value changed from `old` to `new`,
/// matched by key, one per line, in document order of `new`. Panics if
/// the trees' keys differ: two renders of one program always have the
/// same keys.
pub fn compare(old: &Tree, new: &Tree) -> String {
    let (mut before, mut after) = (BTreeMap::new(), BTreeMap::new());
    parts(&old.root, old.root.component.clone(), &mut before);
    parts(&new.root, new.root.component.clone(), &mut after);
    assert_eq!(
        before.keys().collect::<Vec<_>>(),
        after.keys().collect::<Vec<_>>(),
        "two renders of one program have the same keys"
    );
    let mut changes: Vec<(String, String)> = Vec::new();
    for (key, part) in &after {
        match (&before[key], part) {
            (Part::Node { props: a, .. }, Part::Node { at, props: b }) => {
                for name in a
                    .keys()
                    .chain(b.keys())
                    .collect::<std::collections::BTreeSet<_>>()
                {
                    let (x, y) = (a.get(name), b.get(name));
                    if x != y {
                        let show = |v: Option<&serde_json::Value>| {
                            v.map_or("(absent)".to_string(), ToString::to_string)
                        };
                        changes.push((
                            at.clone(),
                            format!("{at} {name}: {} -> {}", show(x), show(y)),
                        ));
                    }
                }
            }
            (Part::Text { text: a, .. }, Part::Text { at, text: b }) if a != b => {
                changes.push((
                    at.clone(),
                    format!(
                        "{at}: {} -> {}",
                        serde_json::to_string(a).unwrap(),
                        serde_json::to_string(b).unwrap()
                    ),
                ));
            }
            (Part::Text { .. }, Part::Text { .. }) => {}
            _ => panic!("a key changed kind"),
        }
    }
    changes.sort();
    changes.into_iter().map(|(_, line)| line + "\n").collect()
}
